### PR TITLE
code clean: unify thread resume code

### PR DESCRIPTION
**Pull request by GangSecurity**

### CODE REVIEW

File: core/arch/x86_64/include/kernel/thread_defs.h [hunk 1]
```diff
@@ -9,5 +9,6 @@
 #define THREAD_FLAGS_COPY_ARGS_ON_RETURN	(1 << 0)
 #define THREAD_FLAGS_FOREIGN_INTR_ENABLE	(1 << 1)
 #define THREAD_FLAGS_EXIT_ON_FOREIGN_INTR	(1 << 2)
+#define THREAD_FLAGS_NEW			(1 << 3)
```
Comment (jingdlu) [inline core/arch/x86_64/include/kernel/thread_defs.h]: Why does Arm not need this flag?

Comment (GangSecurity) [reply core/arch/x86_64/include/kernel/thread_defs.h]: removed.

**Review by xyzhao2018 — COMMENTED** [review]
Please also update the following with macro SAVE_REGS and RESTORE_REGS in exceptions.S.
interrupt_common:
    /* save general purpose registers */
    /* restore general purpose registers */

File: core/arch/x86_64/kernel/thread_64.S [hunk 2]
```diff
@@ -49,32 +49,23 @@
     popfq
 .endm
 
-/* void thread_resume(struct thread_ctx_regs *regs, struct thread_smc_args *args,
- *                      vaddr_t tmp_stack)
+/* void thread_resume(vaddr_t sp, struct thread_smc_args *args,
```
Comment (jingdlu) [inline core/arch/x86_64/kernel/thread_64.S]: Can we just keep one parameter struct thread_ctx_regs *regs for thread_resume? Let's try to get some ideas from Arm code.

Comment (GangSecurity) [reply core/arch/x86_64/kernel/thread_64.S]: No. thread_ctx_regs  is only for registers. In arm design, thread_resume never returns. But for x86, it needs struct thread_smc_args *args for return values. x86 also needs tmp_stack to store current stack.

**Review by xyzhao2018 — APPROVED** [review]

File: core/arch/x86_64/kernel/thread_64.S [hunk 3]
```diff
@@ -15,184 +15,73 @@
...
-    movq %r10, (%rsi)
-    movq %r11, 8(%rsi)
-    movq %r12, 16(%rsi)
-    movq %r13, 24(%rsi)
-    movq %r14, 32(%rsi)
-    movq %r15, 40(%rsi)
+    movq %r15, %rsi   //prepare parameters for thread_smc_args_retrieve 
+    movq $0,   %rdx
+    movq $0,   %rcx
+    movq $0,   %r8
```
Comment (jingdlu) [inline core/arch/x86_64/kernel/thread_64.S]: Should we use xorq here like "xorq %rdx, %rdx ......" ?

Comment (GangSecurity) [reply core/arch/x86_64/kernel/thread_64.S]: updated

**Review by jingdlu — APPROVED** [review]

**Pull request merged**